src/request-data/resources.js

```diff
@@ -9,22 +9,132 @@ https://www.apache.org/licenses/LICENSE-2.0. No part of ODK Central,
     /* eslint-disable no-param-reassign */
     transformResponse: ({ data }) => {
       data.verbs = new Set(data.verbs);
       data.can = hasVerbs;
+      data.preferences = {
+        site: new Proxy(
+          shallowReactive(data.preferences.site),
+          {
+            deleteProperty(target, prop) {
+              const retval = (delete target[prop]);
+              self.preferenceOps.propagate(prop, null, null); // DELETE to backend
+              return retval;
+            },
+            set(target, prop, value) {
+              // eslint-disable-next-line no-multi-assign
+              const retval = (target[prop] = value);
+              self.preferenceOps.propagate(prop, value, null); // PUT to backend
+              return retval;
+            },
+          },
+        ),
+        projects: new Proxy(
+          data.preferences.projects,
+          {
+            deleteProperty() {
+              throw new Error('Deleting a project\'s whole property collection is not supported. Delete each property individually, eg "delete preferences.projects[3].foo".');
+            },
+            set() {
+              throw new Error('Directly setting a project\'s whole property collection is not supported. Set each property individually, eg "preferences.projects[3].foo = \'bar\'"');
+            },
+            get(target, projectId) {
+              if (Number.isNaN(parseInt(projectId, 10))) throw new TypeError(`Not an integer project ID: "${projectId}"`);
+              const projectProps = target[projectId];
+              if (projectProps === undefined || (!isReactive(projectProps))) { // not reentrant (TOCTOU issue) but there's no real way to solve it — as this is supposed to be a synchronous method we can't simply wrap it in a Lock
+                target[projectId] = new Proxy(
+                  // make (potentially autovivicated) props reactive, and front them with a proxy to enable our setters/deleters
+                  shallowReactive(projectProps === undefined ? {} : projectProps),
+                  {
+                    deleteProperty(from, prop) {
+                      const retval = (delete from[prop]);
+                      self.preferenceOps.propagate(prop, null, projectId); // DELETE to backend
+                      return retval;
+                    },
+                    set(from, prop, propval) {
+                      // eslint-disable-next-line no-multi-assign
+                      const retval = (from[prop] = propval);
+                      self.preferenceOps.propagate(prop, propval, projectId); // PUT to backend
+                      return retval;
+                    },
+                  }
+                );
+              }
+              return target[projectId];
+            },
+          }
+        ),
+      };
       return shallowReactive(data);
+    },
+    preferenceOps: {
+      self,
+      _container,
+      abortControllers: {},
+      instanceID: crypto.randomUUID(),
```

Comment:
Yeah, users should be using HTTPS given how Central is deployed. We don't use HTTPS locally, but I think many things still work over HTTP on localhost (e.g., the `Secure` attribute of cookies I think).

Reply:
I can't think of other secure-context-only API incantations, but they may exist. I think it's totally fair to assume it's a secure context. We aren't building Central Frontend to be deployed outside a secure context. If someone does so, they'll need to modify the code. I'm not even sure it's worth the extra time/code to add a warning.

Reply:
We don't offer a lot of support for folks doing custom modifications of Central, which this would be, so I don't think it would lead to a support burden. It's not trivial to get Central Frontend working over HTTP, because the session cookie isn't just `Secure`, it also has a `__Host-` prefix, restricting it to HTTPS. We do some contortions in local development to get the cookie working (main.nginx.conf in this repo).

It's not 100% clear to me that we'll end up needing this function after some of the other changes we've discussed are in. That said, I've thought about using this function for other things before, so I'm glad we're discussing it.